File: packages/xcon/src/evaluator.ts
const INT_RE = /^-?\d+$/;
const FLOAT_RE = /^-?\d+\.\d+$/;

/**
 * Type-infer a bare (unquoted) scalar's text into a JSON-compatible value.
 * Quoted strings must NOT be passed here — they are always strings.
 */
export function inferType(raw: string): string | number | boolean | null {
  if (raw === 'null') return null;
  if (raw === 'true') return true;
  if (raw === 'false') return false;

  if (INT_RE.test(raw)) {
    // Use Number for safe range; fall back to the original string if not safe.
    const n = Number(raw);
    if (Number.isSafeInteger(n)) {
      // Normalize -0 to 0 (JSON does not distinguish them as integers).
      return n === 0 ? 0 : n;
    }
    // Beyond safe integer range — preserve the string to avoid lossy conversion.
    return raw;
  }

  if (FLOAT_RE.test(raw)) {
    return parseFloat(raw);
  }

  return raw;
}

File: packages/xcon-python/xcon/evaluator.py
import re
from typing import Union

_INT_RE = re.compile(r"^-?\d+$")
_FLOAT_RE = re.compile(r"^-?\d+\.\d+$")


def infer_type(raw: str) -> Union[str, int, float, bool, None]:
    """Infer the type of a raw, *bare* (unquoted) value string.

    Quoted strings must NOT pass through this function — they are always strings.
    """
    if raw == "null":
        return None
    if raw == "true":
        return True
    if raw == "false":
        return False
    if _INT_RE.fullmatch(raw):
        n = int(raw)
        # Normalize -0 to 0
        return 0 if n == 0 else n
    if _FLOAT_RE.fullmatch(raw):
        n_f = float(raw)
        return 0.0 if n_f == 0 else n_f
    return raw
